### PR TITLE
NFC user dict list, delete, and de-duplication.

### PR DESCRIPTION
**Pull request by djsime1**
# What's new

- Mifare Classic user keys can be listed and deleted.
- Duplicate keys (from user DB) can't be added to user database.
- A few minor changes to icon names.

# Verification 

- Open Mifare Classic Keys menu in NFC
- If user dict is not empty, list button (right) will be active.
- Add a few keys, then select one in the list to delete.
- Verify the correct key was deleted.
- Try to add a key already present in user database, and get informed it already exists.

# Checklist (For Reviewer)

- [x] PR has description of feature/bug or link to Confluence/Jira task
- [x] Description contains actions to verify feature/bugfix
- [x] I've built this code, uploaded it to the device and verified feature/bugfix

### CODE REVIEW

File: applications/nfc/scenes/nfc_scene_mf_classic_keys_list.c
Comment (djsime1) [inline applications/nfc/scenes/nfc_scene_mf_classic_keys_list.c]: Memory leak was mostly fixed with latest commit, however not completely; I suspect it has to do with this array iterator.

**Comment by gornekich** [comment]
Hello @djsime1. Thank you for the PR!

Dealing with duplicated keys is a good idea. Also our UX/UI team liked your idea with User dictionary keys listing and deleting. However, we want to add this a little bit differently. Our designer is working on UI for this feature right now and soon it will be in public Miro https://miro.com/app/board/uXjVOlznagk=/?share_link_id=903187096944

I will let you know when the design is ready. Please tell me if you want to rework this PR once design is finalized.

**Comment by gornekich** [comment]
Also some changes came to dev recently. Please, sync  the PR with latest dev

**Comment by djsime1** [comment]
Got dev merged, I have limited availability the next few weeks but I'll keep an eye out for that UI/UX being published.

**Comment by djsime1** [comment]
Jumped up to latest dev and fixed issues caused by icon asset changes.

**Comment by gornekich** [comment]
Hello @djsime1 
We updated user dictionary flow here https://miro.com/app/board/uXjVOlznagk=/
Please, tell me if you will implement it in this PR

**Comment by djsime1** [comment]
> Hello @djsime1 We updated user dictionary flow here https://miro.com/app/board/uXjVOlznagk=/ Please, tell me if you will implement it in this PR

Yes, I can implement that flow. I'll go with the top "Key exists" screen since I don't know how to draw a canvas atop the byteinput. Even if I did, switching to a popup scene is still simpler.

image

**Comment by djsime1** [comment]
Hopefully I'll have this ready by tomorrow night.

**Comment by djsime1** [comment]
@gornekich I'm almost ready to push my changes, however Miro won't let me download this icon in its raw format. Could you please reply with the icon attached? Thanks.
image

**Comment by gornekich** [comment]
@djsime1 I added this asset myself, it's already in dev. 
Is your PR ready for review?

**Comment by djsime1** [comment]
> @djsime1 I added this asset myself, it's already in dev. Is your PR ready for review?

The [keychain icon currently present in dev](https://github.com/flipperdevices/flipperzero-firmware/blob/dev/assets/icons/NFC/Keychain.png) is not the same as the Miro keychain icon. The dev one is significantly larger, resulting in the scenes title being truncated from "Mifare Classic Keys" to "MF Classic Keys." ~~Additionally, I think there's one tiny bug about pressing LEFT in the delete confirmation screen.~~

Besides that, it is ready for review. My next commit will ~~fix that LEFT bug,~~ as well as the smaller keychain icon if I receive it before the commit is pushed.

_EDIT: Forgot I already fixed the bug with previous commit, so the icon is the only thing that's left._ **Ready for review!**

**Comment by Karator** [comment]
Hi bro!
We attach that new picture here.
Small_keychain_36x39

It was pinned to Miiro board but sometimes it's tricky to download it ((

**Comment by djsime1** [comment]
Wrong again I was! The confirmation cancel bug was an odd quirk with the order in which scene loading is handled. Long story short, MFC dict was allocated before the if statement, and free'd at the end of the event callback function. Little did I know the current scenes exit and next scenes enter functions were called in between. Sure enough, the next scene's MFC dict allocate statement was called before our free statement. One last commit and this PR will be finished.

**Comment by djsime1** [comment]
Alright, everything's ready to go!

File: applications/nfc/scenes/nfc_scene_mf_classic_keys_list.c
Comment (gornekich) [inline applications/nfc/scenes/nfc_scene_mf_classic_keys_list.c]: Memory leak here. Must be free(*MfClassicUserKeys_ref(it));

File: applications/nfc/scenes/nfc_scene_mf_classic_keys_list.c
Comment (gornekich) [inline applications/nfc/scenes/nfc_scene_mf_classic_keys_list.c]: Really don't like to see static variable here. I think that you can add this array instance in Nfc structure in nfc_i.h, so that it will be dynamically  allocated along with nfc control structure

**Review by gornekich — CHANGES_REQUESTED** [review]
@djsime1 you made a great work, thank you!
Could you please add a few changed and we will merge you PR

**Comment by djsime1** [comment]
Sure, I'll get to work on that when I can.

**Comment by djsime1** [comment]
Could someone please review/merge this commit? I've not gotten any feedback in the past week. Thanks!

**Comment by skotopes** [comment]
I’ll take a look today. We’ve been waiting for elf support to be merged first.

File: applications/main/nfc/scenes/nfc_scene_delete.c
Comment (gornekich) [inline applications/main/nfc/scenes/nfc_scene_delete.c]: This breaks current logic. We want to switch to NfcSceneFileSelect after delete success.

File: applications/main/nfc/scenes/nfc_scene_delete_success.c
Comment (gornekich) [inline applications/main/nfc/scenes/nfc_scene_delete_success.c]: I suggest different approach here. You search for previous scene in scene manager stack and choose where to go next. Something like that:
if(scene_manager_has_previous_scene(nfc->scene_manager, NfcSceneMfClassicKeysAdd) {
    scene_manager_search_and_switch_to_previous_scene(nfc->scene_manager, NfcSceneMfClassicKeys)
} else {
    scene_manager_search_and_switch_to_previous_scene(
                nfc->scene_manager, NfcSceneFileSelect);
}

File: applications/main/nfc/scenes/nfc_scene_dict_not_found.c
Comment (gornekich) [inline applications/main/nfc/scenes/nfc_scene_dict_not_found.c]: Same here. I think it's not good when one scene state is changed from another scene. It's not how scene states are designed to be used

File: applications/main/nfc/scenes/nfc_scene_mf_classic_keys.c
Comment (gornekich) [inline applications/main/nfc/scenes/nfc_scene_mf_classic_keys.c]: There is no need in saving number of user keys in scene state. Once you get user_dict_keys_total in _on_enter() handler, you add button element to widget. If you user_dict_keys_total = 0 and you don't add button element, you will never get GuiButtonRight event

File: applications/main/nfc/scenes/nfc_scene_mf_classic_keys.c
Comment (gornekich) [inline applications/main/nfc/scenes/nfc_scene_mf_classic_keys.c]: Not necessary

**Review by gornekich — CHANGES_REQUESTED** [review]
Please, also revert changes in readme files. Current changes seem strange.
Also, since you changed icon names, our build system increased API version. We don't want to increment major API version for such small changes for now, could you please change API version back to 1.0 by hand. This can be done in api_symbols.csv file.

You can also give maintainers access to push to you branch so we can help to resolve some issues, if you don't mind

**Comment by djsime1** [comment]
Readme changes must have been added by the merge, since I hadn't touched the file at all. I was confused at how the API symbols system worked so I just let it do its thing, but I'll change it back to 1.0. Maintainer push access has been enabled since I originally opened the PR. Lastly, my though process for popup scenes was that you'd be able to dictate where they return too if needed instead of always going to file select. I guess I can undo my changes to the deleted popup since a MFC-Key specific delete scene was added.

File: applications/main/nfc/scenes/nfc_scene_mf_classic_keys.c
Comment (djsime1) [reply applications/main/nfc/scenes/nfc_scene_mf_classic_keys.c]: I didn't know the event only fired when the button was present.

**Comment by djsime1** [comment]
Done.

**Review by gornekich — DISMISSED** [review]
Thanks!

**Review by skotopes — APPROVED** [review]

**Pull request merged**
